pilot/pkg/xds/endpoints/endpoint_builder.go

```diff
@@ -65,67 +70,107 @@ type EndpointBuilder struct {
-	}
-	svc := push.ServiceForHostname(proxy, hostname)
 
+	svc := push.ServiceForHostname(proxy, hostname)
 	var dr *model.ConsolidatedDestRule
 	if svc != nil {
 		dr = proxy.SidecarScope.DestinationRule(model.TrafficDirectionOutbound, proxy, svc.Hostname)
 	}
+
+	return NewCDSEndpointBuilder(
+		proxy, push, clusterName,
+		dir, subsetName, hostname, port,
+		svc, dr,
+	)
+}
+
+// NewCDSEndpointBuilder allows setting some fields directly when we already
+// have the Service and DestinationRule.
+func NewCDSEndpointBuilder(
+	proxy *model.Proxy, push *model.PushContext, clusterName string,
+	dir model.TrafficDirection, subsetName string, hostname host.Name, port int,
+	service *model.Service, dr *model.ConsolidatedDestRule,
+) EndpointBuilder {
 	b := EndpointBuilder{
 		clusterName:     clusterName,
 		network:         proxy.Metadata.Network,
 		proxyView:       proxy.GetView(),
 		clusterID:       proxy.Metadata.ClusterID,
 		locality:        proxy.Locality,
-		service:         svc,
-		clusterLocal:    push.IsClusterLocal(svc),
 		destinationRule: dr,
+		service:         service,
+		clusterLocal:    push.IsClusterLocal(service),
 		nodeType:        proxy.Type,
 
-		mtlsChecker: newMtlsChecker(push, port, dr.GetRule(), subsetName),
-		push:        push,
-		proxy:       proxy,
-		subsetName:  subsetName,
-		hostname:    hostname,
-		port:        port,
-		dir:         dir,
+		subsetName: subsetName,
+		hostname:   hostname,
+		port:       port,
+		push:       push,
+		proxy:      proxy,
+		dir:        dir,
 	}
-
+	b.populateSubsetInfo()
 	b.populateFailoverPriorityLabels()
-
 	return b
 }
 
-// NewTestEndpointBuilder allows setting some fields directly without needing push context.
-// For now, this is for tests. We can expand this to be an all-field constructor for cases
-// like CDS where we already have the service and destination rule.
-func NewTestEndpointBuilder(clusterName string, service *model.Service, dr *model.ConsolidatedDestRule) EndpointBuilder {
-	return EndpointBuilder{
-		clusterName:     clusterName,
-		service:         service,
-		destinationRule: dr,
+func (b *EndpointBuilder) servicePort(port int) *model.Port {
+	if !b.ServiceFound() {
+		log.Debugf("can not find the service %s for cluster %s", b.hostname, b.clusterName)
+		return nil
+	}
+	svcPort, f := b.service.Ports.GetByPort(port)
+	if !f {
+		log.Debugf("can not find the service port %d for cluster %s", b.port, b.clusterName)
+		return nil
+	}
+	return svcPort
+}
+
+func (b *EndpointBuilder) WithSubset(subset string) EndpointBuilder {
+	subsetBuilder := *b
+	subsetBuilder.subsetName = subset
+	subsetBuilder.populateSubsetInfo()
+	return subsetBuilder
+}
+
+func (b *EndpointBuilder) populateSubsetInfo() {
+	if b.dir == model.TrafficDirectionInboundVIP {
+		b.subsetName = strings.TrimPrefix(b.subsetName, "http/")
+		b.subsetName = strings.TrimPrefix(b.subsetName, "tcp/")
+	}
+	b.mtlsChecker = newMtlsChecker(b.push, b.port, b.destinationRule.GetRule(), b.subsetName)
```

Comment:
nit: donot couple mtlsChecker with subsetInfo

Reply:
mtlsChecker depends on the subset name. I could make it so we build it in the constructor, and the build it again if the subset changes?